client/rpcproxy/rpcproxy.go

```diff
@@ -0,0 +1,765 @@
+
+	p.rebalanceTimer.Reset(connRebalanceTimeout)
+	return connRebalanceTimeout
+}
+
+// ResetRebalanceTimer resets the rebalance timer.  This method exists for
+// testing and should not be used directly.
+func (p *RpcProxy) ResetRebalanceTimer() {
+	p.listLock.Lock()
+	defer p.listLock.Unlock()
+	p.rebalanceTimer.Reset(clientRPCMinReuseDuration)
+}
+
+// ServerRPCAddrs returns one RPC Address per server
+func (p *RpcProxy) ServerRPCAddrs() []string {
+	l := p.getServerList()
+	serverAddrs := make([]string, 0, len(l.L))
+	for _, s := range l.L {
+		serverAddrs = append(serverAddrs, s.Addr.String())
+	}
+	return serverAddrs
+}
+
+// Run is used to start and manage the task of automatically shuffling and
+// rebalancing the list of Nomad servers.  This maintenance only happens
+// periodically based on the expiration of the timer.  Failed servers are
+// automatically cycled to the end of the list.  New servers are appended to
+// the list.  The order of the server list must be shuffled periodically to
+// distribute load across all known and available Nomad servers.
+func (p *RpcProxy) Run() {
+	for {
+		select {
+		case <-p.rebalanceTimer.C:
+			p.RebalanceServers()
+
+			p.refreshServerRebalanceTimer()
+		case <-p.shutdownCh:
+			p.logger.Printf("[INFO] RPC Proxy: shutting down")
+			return
+		}
+	}
+}
+
+// UpdateFromNodeUpdateResponse handles heartbeat responses from Nomad
+// Servers.  Heartbeats contain a list of Nomad Servers that the client
+// should talk with for RPC requests.  UpdateFromNodeUpdateResponse does not
+// rebalance its serverList, that is handled elsewhere.  New servers learned
+// via the heartbeat are appended to the RpcProxy's serverList.  Removed
+// servers are removed immediately.  Servers speaking a newer RPC version are
+// filtered from the serverList.
+func (p *RpcProxy) UpdateFromNodeUpdateResponse(resp *structs.NodeUpdateResponse) error {
+	// Merge all servers found in the response.  Servers in the response
+	// with newer API versions are filtered from the list.  If the list
+	// is missing an address found in the RpcProxy's server list, remove
+	// it from the RpcProxy.
+	//
+	// FIXME(sean@): This is not true.  We rely on an outside pump to set
+	// these values.  In order to catch the orphaned clients where all
+	// Nomad servers were rolled between the heartbeat interval, the
+	// rebalance task queries Consul and adds the servers found in Consul
+	// to the server list in order to reattach an orphan to a server.
+
+	p.serverListLock.Lock()
+	defer p.serverListLock.Unlock()
+
+	// Clear the backup server list when a heartbeat contains at least
+	// one server.
+	if len(resp.Servers) > 0 && len(p.backupServers.L) > 0 {
+		p.backupServers.L = make([]*ServerEndpoint, 0, len(resp.Servers))
+	}
+
+	// 1) Create a map to reconcile the difference between
+	// m.primaryServers and resp.Servers.
+	type targetServer struct {
```

Comment:
I would like to see a function that takes these two server lists and returns three sets: added, removed and same. Then we can dedup this set difference logic